### ISSUE TITLE
Type check for lambda fails with NamedTuple

### ISSUE THREAD

**Issue by Zaikaba** [microsoft/pyright#9110]
## Environment data

-   Language Server version: 2024.9.2
-   OS and version: win32 x64
-   Python version (and distribution if applicable, e.g. Anaconda): 3.12.5 ( conda )
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: strict

## Code Snippet

```python
from collections.abc import Callable
from typing import NamedTuple


class Bar1:

    def __init__( self, val: bool, fun: Callable[ [ bool ], bool ] ):
        self.val = val
        self.fun = fun

# fine: "x: bool"
bar1 = Bar1( True, lambda x: x )

class Bar2( NamedTuple ):

    val: bool
    fun: Callable[ [ bool ], bool ]


# Type of parameter "x" is unknown
bar = Bar2( True, lambda x: x )

```

![error](https://github.com/user-attachments/assets/21879a0f-4eda-4101-bcb0-f1ad2a3cfa84)

## Expected behavior

Type check should derive (bool)->bool in both cases.

## Actual behavior

NamedTuple version fails, x is unknown.

## Logs

```

```

**Comment by erictraut** [comment]
This is an unfortunate consequence of the fact that the `NamedTuple` runtime support synthesizes a `__new__` method with a detailed signature, but the `__init__ method for a `NamedTuple` class has simple `*args: Any` and `**kwargs: Any` parameters.

Updating your comparison with `Bar1` to match this behavior of a `NamedTuple`, we can see that pyright's behavior is consistent here.

```python
class Bar1:
    def __new__(cls, val: bool, fun: Callable[[bool], bool]) -> Self: ...
    def __init__(self, *args: Any, **kwargs: Any) -> None: ...

bar1 = Bar1(True, lambda x: x) # Type of parameter "x" is unknown
```

Pyright evaluates the argument expressions for both the `__new__` and the `__init__` method for classes in conformance with [the typing spec](https://typing.readthedocs.io/en/latest/spec/constructors.html#constructor-calls). The final type of the argument expressions is dictated by the `__init__` method. In this case, the `__init__` method parameter types do not allow for the lambda to be evaluated with types more accurate than `Unknown`.

I presume that you have strict-mode type checking enabled or have specifically enabled the `reportUnknownLambdaType` check. If you don't want to see these errors, you can disable this check. If you want to leave it enabled, you'll need to suppress it in this case or modify your code to avoid this condition. One option is to swap out your named tuple for a frozen dataclass, which has similar semantics.

In any case, pyright is working as intended here, so I don't consider this a bug.

**Comment by A5rocks** [comment]
> Pyright evaluates the argument expressions for both the `__new__` and the `__init__` method for classes in conformance with [the typing spec](https://typing.readthedocs.io/en/latest/spec/constructors.html#constructor-calls). The final type of the argument expressions is dictated by the `__init__` method. In this case, the `__init__` method parameter types do not allow for the lambda to be evaluated with types more accurate than `Unknown`.

I don't see why this has to be the case? I understand that it might be annoying if `__new__` and `__init__` specify different types and doing type inference for one creates issues for the other, but surely pyright can special case if one is `Unknown`? (I wrote something about intersections but actually I don't know, so edited) I think given 3 (?) people have run into this so it's probably a good idea?
